src/iface/fragments.rs

```diff
@@ -0,0 +1,278 @@
+                        if front == *total_len {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Retrieve the first continuous segment from the assembler
+    pub fn front(&mut self) -> Option<usize> {
+        match self.state {
+            PacketState::Empty => {
+                panic!("Packet is empty, cannot return front");
+            },
+            PacketState::Assembling{ref mut assembler, ..} => {
+                assembler.remove_front()
+            }
+        }
+    }
+
+    /// Is the packet empty?
+    pub fn is_empty(&self) -> bool {
+        match self.state {
+            PacketState::Empty => {
+                true
+            },
+            PacketState::Assembling{..} => {
+                false
+            }
+        }
+    }
+
+    /// Initiate reassembly of a frahgmented packet
+    pub fn start(&mut self, id: u16, src_addr: Ipv4Address, dst_addr: Ipv4Address) {
+        match self.state {
+            PacketState::Empty => {
+                self.state = PacketState::Assembling {
+                    assembler: Assembler::new(self.rx_buffer.len()),
+                    id: id,
+                    src_addr: src_addr,
+                    dst_addr: dst_addr,
+                    last_used: Instant::from_millis(0),
+                    has_header: false,
+                    header_len: 0,
+                    total_len: 0,
+                }
+            },
+            PacketState::Assembling{..} => {
+                panic!("Attempting to start an assembling packet");
+            }
+        }
+    }
+
+    /// Set total length of the reassembled packet
+    pub fn set_total_len(&mut self, len: usize) {
+        match self.state {
+            PacketState::Empty => {
+                panic!("Packet is empty, cannot set total length of the fragment");
+            },
+            PacketState::Assembling{ref mut total_len, ..} => {
+                *total_len = len;
+            }
+        }
+    }
+
+    /// Get an immutable slice of the underlying packet data
+    pub fn get_buffer(&self, start: usize, end: usize) -> &[u8] {
+        match self.state {
+            PacketState::Empty => {
+                panic!("Packet is empty, cannot access its buffer");
+            },
+            PacketState::Assembling{..} => {
+                &self.rx_buffer[start..end]
+            }
+        }
+    }
+
+    /// Get a mutable slice of the underlying packet data
+    pub fn get_buffer_mut(&mut self, start: usize, end: usize) -> &mut [u8] {
+        match self.state {
+            PacketState::Empty => {
+                panic!("Packet is empty, cannot access its buffer");
+            },
+            PacketState::Assembling{..} => {
+                &mut self.rx_buffer[start..end]
+            }
+        }
+    }
+}
+
+pub struct Set<'a> {
+	packets: ManagedSlice<'a, Packet<'a>>,
```

Comment:
I would like this to use a ManagedMap as I have described previously.

Reply:
With ManagedMap and std it goes as follows:

1. new fragment arrives
2. fragment is not in the FragmentSet, create a new FragmentedPacket (its storage backed by a `vec!`) and insert it with the key to the FragmentSet

Now, with `no_std`, you cannot create a new FragmentedPacket with a `vec!` - you need an array, and you cannot allocate it from within the `process_ipv4_fragment` function because of lifetimes. So all FragmentedPackets have to be initialized prior polling (hence ManagedSlice).

Does it make sense?